File: examples/node/apps/agent/mcp_http.ts
import {
  AppMcpProtocolError,
  BaseAppMcpHttpAdapter,
  type AppMcpCallResult,
  type AppMcpHttpExchange,
  type AppMcpHttpResponse,
  type AppMcpInitializeResult,
  type AppMcpReadResourceResult,
  type AppMcpRequestContext,
  type AppMcpResourceDescriptor,
  type AppMcpServer,
  type AppMcpToolDescriptor,
} from "../../core/shared/app_mcp_contracts";

type JsonRpcId = string | number | null;

type JsonRpcRequest = {
  jsonrpc: "2.0";
  id: JsonRpcId;
  method: string;
  params?: unknown;
};

type JsonRpcNotification = {
  jsonrpc: "2.0";
  method: string;
  params?: unknown;
};

type JsonRpcSuccess = {
  jsonrpc: "2.0";
  id: JsonRpcId;
  result:
    | AppMcpInitializeResult
    | { tools: AppMcpToolDescriptor[] }
    | { resources: AppMcpResourceDescriptor[] }
    | AppMcpReadResourceResult
    | AppMcpCallResult
    | Record<string, never>;
};

type JsonRpcFailure = {
  jsonrpc: "2.0";
  id: JsonRpcId;
  error: {
    code: number;
    message: string;
    data?: unknown;
  };
};

function generateId(): string {
  return globalThis.crypto?.randomUUID?.() ?? Math.random().toString(36).slice(2);
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function toFailure(
  id: JsonRpcId,
  code: number,
  message: string,
  data?: unknown
): JsonRpcFailure {
  return {
    jsonrpc: "2.0",
    id,
    error: {
      code,
      message,
      ...(data !== undefined ? { data } : {}),
    },
  };
}

function isJsonRpcMessage(value: unknown): value is {
  jsonrpc: "2.0";
  method: string;
  id?: JsonRpcId;
  params?: unknown;
} {
  return (
    isObject(value) &&
    value.jsonrpc === "2.0" &&
    typeof value.method === "string"
  );
}

function isInitializeParams(value: unknown): value is {
  protocolVersion: string;
  capabilities?: Record<string, unknown>;
  clientInfo?: {
    name: string;
    version: string;
  };
} {
  if (!isObject(value) || typeof value.protocolVersion !== "string") {
    return false;
  }

  if (value.clientInfo === undefined) {
    return true;
  }

  return (
    isObject(value.clientInfo) &&
    typeof value.clientInfo.name === "string" &&
    typeof value.clientInfo.version === "string"
  );
}

function toJsonResponse(
  statusCode: number,
  bodyText?: string,
  headers?: Record<string, string>
): AppMcpHttpResponse {
  return {
    statusCode,
    headers: {
      "cache-control": "no-store",
      ...(bodyText !== undefined
        ? { "content-type": "application/json; charset=utf-8" }
        : {}),
      ...headers,
    },
    bodyText,
  };
}

export class StreamableHttpAppMcpAdapter extends BaseAppMcpHttpAdapter {
  public readonly transport = "streamable-http" as const;
  public readonly endpointPath = "/mcp" as const;

  public async handle(
    exchange: AppMcpHttpExchange,
    server: AppMcpServer
  ): Promise<AppMcpHttpResponse | undefined> {
    if (exchange.path !== this.endpointPath) {
      return undefined;
    }

    const method = exchange.method.toUpperCase();

    if (method === "GET") {
      return {
        statusCode: 405,
        headers: {
          allow: "POST",
          "cache-control": "no-store",
        },
      };
    }

    if (method === "DELETE") {
      return {
        statusCode: 405,
        headers: {
          allow: "POST",
          "cache-control": "no-store",
        },
      };
    }

    if (method !== "POST") {
      return {
        statusCode: 405,
        headers: {
          allow: "GET,POST,DELETE",
          "cache-control": "no-store",
        },
      };
    }

    if (!exchange.bodyText?.trim()) {
      return toJsonResponse(
        400,
        JSON.stringify(toFailure(null, -32600, "Missing JSON-RPC payload."))
      );
    }

    let parsed: unknown;
    try {
      parsed = JSON.parse(exchange.bodyText) as unknown;
    } catch (error: unknown) {
      return toJsonResponse(
        400,
        JSON.stringify(
          toFailure(
            null,
            -32700,
            "Invalid JSON-RPC payload.",
            error instanceof Error ? error.message : undefined
          )
        )
      );
    }

    if (Array.isArray(parsed) && parsed.length === 0) {
      return toJsonResponse(
        400,
        JSON.stringify(
          toFailure(null, -32600, "Empty JSON-RPC batch payload is invalid.")
        )
      );
    }

    const messages = Array.isArray(parsed) ? parsed : [parsed];
    const responses: Array<JsonRpcSuccess | JsonRpcFailure> = [];

    for (const message of messages) {
      if (!isJsonRpcMessage(message)) {
        responses.push(
          toFailure(null, -32600, "Invalid JSON-RPC request shape.")
        );
        continue;
      }

      const hasId = Object.prototype.hasOwnProperty.call(message, "id");
      if (!hasId) {
        this.handleNotification(message);
        continue;
      }

      const request: JsonRpcRequest = {
        jsonrpc: "2.0",
        id:
          typeof message.id === "string" || typeof message.id === "number"
            ? message.id
            : null,
        method: message.method,
        params: message.params,
      };

      try {
        responses.push(await this.handleRequest(request, server));
      } catch (error: unknown) {
        if (error instanceof AppMcpProtocolError) {
          responses.push(
            toFailure(request.id, error.code, error.message, error.data)
          );
          continue;
        }

        responses.push(
          toFailure(
            request.id,
            -32603,
            "Internal MCP server error.",
            error instanceof Error ? { message: error.message } : undefined
          )
        );
      }
    }

    if (responses.length === 0) {
      return {
        statusCode: 202,
        headers: {
          "cache-control": "no-store",
        },
      };
    }

    return toJsonResponse(
      200,
      JSON.stringify(responses.length === 1 ? responses[0] : responses)
    );
  }

  private handleNotification(notification: JsonRpcNotification): void {
    switch (notification.method) {
      case "notifications/initialized":
      case "notifications/cancelled":
        return;
      default:
        return;
    }
  }

  private async handleRequest(
    request: JsonRpcRequest,
    server: AppMcpServer
  ): Promise<JsonRpcSuccess> {
    const context = this.toContext(request.id);

    switch (request.method) {
      case "initialize":
        return {
          jsonrpc: "2.0",
          id: request.id,
          result: await server.initialize(
            isInitializeParams(request.params) ? request.params : undefined,
            context
          ),
        };
      case "ping":
        return {
          jsonrpc: "2.0",
          id: request.id,
          result: {},
        };
      case "tools/list":
        return {
          jsonrpc: "2.0",
          id: request.id,
          result: {
            tools: await server.listTools(context),
          },
        };
      case "resources/list":
        return {
          jsonrpc: "2.0",
          id: request.id,
          result: {
            resources: await server.listResources(context),
          },
        };
      case "resources/read":
        if (!isObject(request.params) || typeof request.params.uri !== "string") {
          throw new AppMcpProtocolError(
            -32602,
            "MCP resources/read requires a string resource uri."
          );
        }

        return {
          jsonrpc: "2.0",
          id: request.id,
          result: await server.readResource(request.params.uri, context),
        };
      case "tools/call":
        if (!isObject(request.params) || typeof request.params.name !== "string") {
          throw new AppMcpProtocolError(
            -32602,
            "MCP tools/call requires a string tool name."
          );
        }

        return {
          jsonrpc: "2.0",
          id: request.id,
          result: await server.callTool(
            request.params.name,
            request.params.arguments,
            context
          ),
        };
      default:
        throw new AppMcpProtocolError(
          -32601,
          `MCP method ${request.method} is not implemented by this server.`
        );
    }
  }

  private toContext(requestId?: JsonRpcId): AppMcpRequestContext {
    return {
      transport: this.transport,
      requestId,
      correlationId: generateId(),
    };
  }
}

File: examples/python/apps/agent/mcp_http.py
# ==========================================================================
# APP v1.1.6
# apps/agent/mcp_http.py
# --------------------------------------------------------------------------
# MCP streamable-http transport adapter.
#
# Handles JSON-RPC over HTTP POST at /mcp.
# Supports single and batch requests.
# Returns 405 for GET/DELETE, 202 for notification-only batches.
# ==========================================================================

from __future__ import annotations

import json
from typing import Any

from core.shared.app_mcp_contracts import (
    AppMcpHttpExchange,
    AppMcpHttpResponse,
    AppMcpProtocolError,
    AppMcpRequestContext,
    AppMcpServer,
    BaseAppMcpHttpAdapter,
)

_JSONRPC_VERSION = "2.0"
_PARSE_ERROR = -32700
_INVALID_REQUEST = -32600
_METHOD_NOT_FOUND = -32601
_INTERNAL_ERROR = -32603

_JSON_HEADERS = {"Content-Type": "application/json"}


class StreamableHttpAppMcpAdapter(BaseAppMcpHttpAdapter):
    """MCP adapter for streamable HTTP transport (POST /mcp)."""

    @property
    def transport(self) -> str:
        return "streamable-http"

    @property
    def endpoint_path(self) -> str:
        return "/mcp"

    def handle(
        self,
        exchange: AppMcpHttpExchange,
        server: AppMcpServer,
    ) -> AppMcpHttpResponse | None:
        """Handle an incoming HTTP exchange for MCP."""
        # Only POST is supported.
        if exchange.method.upper() != "POST":
            return AppMcpHttpResponse(
                status_code=405,
                headers={**_JSON_HEADERS, "Allow": "POST"},
                body_text=json.dumps({
                    "jsonrpc": _JSONRPC_VERSION,
                    "id": None,
                    "error": {"code": _INVALID_REQUEST, "message": "Method not allowed. Use POST."},
                }),
            )

        # Parse body
        body_text = exchange.body_text or ""
        try:
            parsed = json.loads(body_text)
        except json.JSONDecodeError:
            return _error_response(None, _PARSE_ERROR, "Parse error")

        # Single vs batch
        if isinstance(parsed, list):
            return self._handle_batch(parsed, server)
        elif isinstance(parsed, dict):
            return self._handle_single(parsed, server)
        else:
            return _error_response(None, _INVALID_REQUEST, "Invalid JSON-RPC request")

    def _handle_single(
        self, msg: dict[str, Any], server: AppMcpServer
    ) -> AppMcpHttpResponse:
        """Handle a single JSON-RPC message."""
        request_id = msg.get("id")
        method = msg.get("method", "")
        params = msg.get("params", {})
        is_notification = request_id is None

        if is_notification:
            self._handle_notification(method, params, server)
            return AppMcpHttpResponse(status_code=202, headers=_JSON_HEADERS)

        return self._handle_request(request_id, method, params, server)

    def _handle_batch(
        self, messages: list[Any], server: AppMcpServer
    ) -> AppMcpHttpResponse:
        """Handle a batch of JSON-RPC messages."""
        responses: list[dict[str, Any]] = []
        has_requests = False

        for msg in messages:
            if not isinstance(msg, dict):
                responses.append({
                    "jsonrpc": _JSONRPC_VERSION,
                    "id": None,
                    "error": {"code": _INVALID_REQUEST, "message": "Invalid request in batch"},
                })
                continue

            request_id = msg.get("id")
            method = msg.get("method", "")
            params = msg.get("params", {})
            is_notification = request_id is None

            if is_notification:
                self._handle_notification(method, params, server)
            else:
                has_requests = True
                resp = self._handle_request(request_id, method, params, server)
                if resp.body_text:
                    responses.append(json.loads(resp.body_text))

        if not has_requests:
            return AppMcpHttpResponse(status_code=202, headers=_JSON_HEADERS)

        return AppMcpHttpResponse(
            status_code=200,
            headers=_JSON_HEADERS,
            body_text=json.dumps(responses, ensure_ascii=False),
        )

    def _handle_notification(
        self, method: str, params: Any, server: AppMcpServer
    ) -> None:
        """Handle a notification (no response needed)."""
        # Silently accept known notifications.
        pass

    def _handle_request(
        self,
        request_id: Any,
        method: str,
        params: Any,
        server: AppMcpServer,
    ) -> AppMcpHttpResponse:
        """Dispatch a JSON-RPC request to the server."""
        ctx = AppMcpRequestContext(transport="streamable-http", request_id=request_id)

        try:
            if method == "initialize":
                result = server.initialize(params, ctx)
                return _success_response(request_id, result)

            if method == "ping":
                return _success_response(request_id, {})

            if method == "tools/list":
                tools = server.list_tools(ctx)
                return _success_response(request_id, {"tools": tools})

            if method == "resources/list":
                resources = server.list_resources(ctx)
                return _success_response(request_id, {"resources": resources})

            if method == "resources/read":
                uri = (params or {}).get("uri", "")
                result = server.read_resource(uri, ctx)
                return _success_response(request_id, result)

            if method == "tools/call":
                name = (params or {}).get("name", "")
                args = (params or {}).get("arguments", {})
                result = server.call_tool(name, args, ctx)
                return _success_response(request_id, result)

            return _error_response(request_id, _METHOD_NOT_FOUND, f"Unknown method: {method}")

        except AppMcpProtocolError as e:
            return _error_response(request_id, e.code, str(e), e.data)
        except Exception as e:
            return _error_response(request_id, _INTERNAL_ERROR, str(e))


# ==========================================================================
# Helpers
# ==========================================================================

def _success_response(request_id: Any, result: Any) -> AppMcpHttpResponse:
    """Build a JSON-RPC success HTTP response."""
    return AppMcpHttpResponse(
        status_code=200,
        headers=_JSON_HEADERS,
        body_text=json.dumps({
            "jsonrpc": _JSONRPC_VERSION,
            "id": request_id,
            "result": result,
        }, ensure_ascii=False),
    )


def _error_response(
    request_id: Any,
    code: int,
    message: str,
    data: Any = None,
) -> AppMcpHttpResponse:
    """Build a JSON-RPC error HTTP response."""
    error: dict[str, Any] = {"code": code, "message": message}
    if data is not None:
        error["data"] = data
    return AppMcpHttpResponse(
        status_code=200,  # JSON-RPC errors use 200 status with error payload.
        headers=_JSON_HEADERS,
        body_text=json.dumps({
            "jsonrpc": _JSONRPC_VERSION,
            "id": request_id,
            "error": error,
        }, ensure_ascii=False),
    )
